### DISCUSSION TITLE
Option optional value doesn't work because `is_flag` is not passed to click properly

### DISCUSSION THREAD

**Discussion by iurisilvio** [fastapi/typer#873] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the Typer documentation, with the integrated search.
- [X] I already searched in Google "How to X in Typer" and didn't find any information.
- [X] I already read and followed all the tutorials in the docs and didn't find an answer.
- [X] I already checked if it is not related to Typer but to [Click](https://github.com/pallets/click).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import typer

def main(
    foo: str = typer.Option(is_flag=False, flag_value="flag", default="default"),
):
    print(f"Hello {foo}")

if __name__ == "__main__":
    typer.run(main)
```

### Description

Click support options with [optional values](https://click.palletsprojects.com/en/8.1.x/options/#optional-value).

I'd expect Typer to work the same way, because it expose both `is_flag` and `flag_value`, but it fails.

I would expect this:

```
$ python foo.py
Hello default
$ python foo.py --foo
Hello flag
$ python foo.py --foo value
Hello value
```

But this is what I get:

```
$ python foo.py
Hello flag
$ python foo.py --foo
Hello flag
$ python foo.py --foo value
Usage: foo.py [OPTIONS]
Try 'foo.py --help' for help.
Error: Got unexpected extra argument (value)
```

It works nice with https://github.com/tiangolo/typer/pull/872.

### Operating System

macOS

### Typer Version

0.12.3

### Python Version

Python 3.11.9

**Comment by tiangolo** [comment, marked as answer]
Thanks for pointing this out! I realized it doesn't make sense to have that parameter when it's not really supported in Typer. And I consider it's better to have only the `boot` type annotation, and if the application code needs a string, then it can check if the `bool` is `True` or `False`, instead of having it as part of the library.

Removing the parameter was done here: https://github.com/fastapi/typer/pull/987

**Comment by alanwilter** [comment]
Heavens, please show an example of how to use your approach and not the `click` one!

I need to handle --my_arg whose default is 'hello':

```sh
app
# my_arg is None or False
app --my_arg
# my_arg = hello
app --my_arg bye
# my_arg = bye
```

**Reply by jnu** [reply]
+1 for this. It's common to distinguish between these three cases (no option passed, option passed with no argument, and option passed with argument). Doesn't seem like there's a way to do it in Typer right now, which is disappointing.
